site/docs/plugins/conversations.md

```diff
@@ -7,1190 +7,1501 @@ next: false
+  const ctx2 = await conversation.wait(); //  B
+  await ctx1.reply("Goodbye!"); //
+} //
+```
+
+```ts [Replay 2]
+async function hello( //                      .
+  conversation: Conversation, //              .
+  ctx0: Context, //                           .
+) { //                                        .
+  await ctx0.reply("Hi there!"); //           .
+  const ctx1 = await conversation.wait(); //  A
+  await ctx1.reply("Hello again!"); //        .
+  const ctx2 = await conversation.wait(); //  B
+  await ctx1.reply("Goodbye!"); //            |
+} //                                          —
 ```
 
 :::
 
-Can you figure out how this bot will work?
+1. When the conversation is entered, the function will run until `A`.
+2. When the next update arrives, the function will be replayed until `A`, and run normally from `A` until `B`.
+3. When the last update arrives, the function will be replayed until `B`, and run normally until the end.
 
-## Installing and Entering a Conversation
+This means that each line of code you write will be executed many times---once normally, and many more times during replays.
+As a result, you have to make sure that your code behaves the same way during replays as it did when it was first executed.
 
-First of all, you **must** use the [session plugin](./session) if you want to use the conversations plugin.
-You also have to install the conversations plugin itself, before you can register individual conversations on your bot.
+If you perform any API calls via `ctx.api` (including `ctx.reply`), the plugin takes care of them automatically.
+In contrast, your own database communcation needs special treatment.
 
-```ts
-// Install the session plugin.
-bot.use(session({
-  initial() {
-    // return empty object for now
-    return {};
-  },
-}));
+This is done as follows.
 
-// Install the conversations plugin.
-bot.use(conversations());
-```
+### The Golden Rule of Conversations
 
-Next, you can install the conversation builder function as middleware on your bot object by wrapping it inside `createConversation`.
+Now that [we know how conversations are executed](#conversations-are-replay-engines), we can define one rule that applies to the code you write inside a conversation builder function.
+You must follow it if you want your code to behave correctly.
 
-```ts
-bot.use(createConversation(greeting));
-```
+::: warning THE GOLDEN RULE
+
+**Code behaving differently between replays must be wrapped in `conversation.external`.**
 
-Now that your conversation is registered on the bot, you can enter the conversation from any handler.
-Make sure to use `await` for all methods on `ctx.conversation`---otherwise your code will break.
+:::
+
+This is how to apply it:
 
 ```ts
-bot.command("start", async (ctx) => {
-  await ctx.conversation.enter("greeting");
-});
+// BAD
+const response = await accessDatabase();
+// GOOD
+const response = await conversation.external(() => accessDatabase());
```

Comment:
Not sure if this is correct since the previous example uses `await` for `accessDatabase()`
```suggestion
// BAD
const response = await accessDatabase();
// GOOD
const response = await conversation.external(async () => await accessDatabase());
```

Reply:
There is still an `await`. All promises are handled correctly.